src/sfizz/EQPool.cpp

```diff
@@ -0,0 +1,137 @@
+#include "EQPool.h"
+#include "AtomicGuard.h"
+#include <thread>
+#include "absl/algorithm/container.h"
+#include "SIMDHelpers.h"
+using namespace std::chrono_literals;
+
+sfz::EQHolder::EQHolder(const MidiState& state)
+:midiState(state)
+{
+    eq.setChannels(2);
+}
+
+void sfz::EQHolder::reset()
+{
+    eq.clear();
+}
+
+void sfz::EQHolder::setup(const EQDescription& description, uint8_t velocity)
+{
+    reset();
+    this->description = &description;
+    const auto normalizedVelocity = normalizeVelocity(velocity);
+
+    baseBandwidth = description.bandwidth;
+    baseGain = Default::eqGainRange.clamp(description.gain + normalizedVelocity * description.vel2gain);
+    baseFrequency = Default::eqFrequencyRange.clamp(description.frequency + normalizedVelocity * description.vel2frequency);
+}
+
+void sfz::EQHolder::process(const float** inputs, float** outputs, unsigned numFrames)
+{
+    if (description == nullptr) {
+        for (unsigned channelIdx = 0; channelIdx < eq.channels(); channelIdx++)
+            copy<float>({ inputs[channelIdx], numFrames }, { outputs[channelIdx], numFrames });
+        return;
+    }
+
+    // TODO: Once the midistate envelopes are done, add modulation in there!
+    // For now we take the last value
+    lastFrequency = baseFrequency;
+    for (auto& mod: description->frequencyCC) {
+        lastFrequency += midiState.getCCValue(mod.first) * mod.second;
+    }
+    lastFrequency = Default::eqFrequencyRange.clamp(lastFrequency);
+
+    lastBandwidth = baseBandwidth;
+    for (auto& mod: description->bandwidthCC) {
+        lastBandwidth += midiState.getCCValue(mod.first) * mod.second;
+    }
+    lastBandwidth = Default::eqBandwidthRange.clamp(lastBandwidth);
+
+    lastGain = baseGain;
+    for (auto& mod: description->gainCC) {
+        lastGain += midiState.getCCValue(mod.first) * mod.second;
+    }
+    lastGain = Default::eqGainRange.clamp(lastGain);
+
+    eq.process(inputs, outputs, lastFrequency, lastBandwidth, lastGain, numFrames);
+}
+float sfz::EQHolder::getLastFrequency() const
+{
+    return lastFrequency;
+}
+float sfz::EQHolder::getLastBandwidth() const
+{
+    return lastBandwidth;
+}
+float sfz::EQHolder::getLastGain() const
+{
+    return lastGain;
+}
+void sfz::EQHolder::setSampleRate(float sampleRate)
+{
+    eq.init(static_cast<double>(sampleRate));
+}
+
+sfz::EQPool::EQPool(const MidiState& state, int numEQs)
+: midiState(state)
+{
+    setnumEQs(numEQs);
+}
+
+sfz::EQHolderPtr sfz::EQPool::getEQ(const EQDescription& description, uint8_t velocity)
+{
+    AtomicGuard guard { givingOutEQs };
```

Comment:
I wonder, what is the purpose of `AtomicGuard`?
does it provide more value that attempting try-lock on a regular mutex?

Reply:
In the way I'd like it to work, not much, but I wasn't sure that locking the mutex if the try-lock succeeds would be realtime-safe on all platforms. Apparently linux uses futexes which should be (?) but I don't know how it all works inside and if it's also the case for MacOS and Windows. Atomics looked a bit safer, although I'm not super fond of the usage of these classes.

If you have better ways I'm actually interested to know more :)

Reply:
No very much detail on it, it's one of many obscure aspects of low level audio programming.
Fwiw, I think it to be generally accepted in the common audio programming wisdom.

Some articles with mentions
http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing
http://atastypixel.com/blog/four-common-mistakes-in-audio-development/

You can find it also in Ardour at various places.
https://github.com/Ardour/ardour/search?q=try_lock&unscoped_q=try_lock

Reply:
If it's good enough for Ardour, it's probably good enough for me.